sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CodegenObjectFactory.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.codehaus.commons.compiler.CompileException
+import org.codehaus.janino.InternalCompilerException
+
+object CodegenObjectFactory {
```

Comment:
Can we make this an abstract class in which we just need to implement hooks for the  code generated and interpreted versions, i.e.:
```scala
object CodegenError {
  def unapply(throwable: Throwable): Option[Exception] = throwable match {
    case e: InternalCompilerException => Some(e)
    case e: CompileException => Some(e)
    case _ => None
  }
}

abstract class CodegenObjectFactory[IN, OUT] {
  def create(in: IN): OUT = try createCodeGeneratedObject(in) catch {
    case CodegenError(_) =>createInterpretedObject(in)
  }

  protected def createCodeGeneratedObject(in: IN): OUT
  protected def createInterpretedObject(in: IN): OUT
}

object UnsafeProjectionFactory extends CodegenObjectFactory[Seq[Expression], UnsafeProjection] {
  ...
}
```

Reply:
This looks good to me, except that I think `UnsafeProjectionCreator` should still support `UnsafeProjectionCreator`'s APIs. So we can replace all usage of `UnsafeProjection` with `UnsafeProjectionCreator`.